File: Python/Queue/Queue.py
class QueueArray:
    def __init__(self, array_size):
        self._front = -1
        self._rear = -1
        self._array = [0] * array_size

    def enqueue(self, value):
        if self.is_empty():
            self._front += 1
        if not self.is_full():
            self._rear += 1
            self._array[self._rear] = value

    def dequeue(self):
        if self.is_empty():
            return None
        target = self._array[self._front]
        self._array[self._front] = 0
        if self._front == self._rear:
            self._front = self._rear = -1
        else:
            self._front += 1
        return target

    def peek(self):
        if self.is_empty():
            return None
        return self._array[self._front]

    def is_empty(self):
        if self._front == -1:
            return True
        return False

    def is_full(self):
        if self._rear == len(self._array):
            return True
        return False


class Node:
    def __init__(self, data):
        self._data = data
        self._previous = None
        self._next = None

    @property
    def data(self):
        return self._data

    @property
    def previous(self):
        return self._previous

    @previous.setter
    def previous(self, node):
        self._previous = node

    @property
    def next(self):
        return self._next

    @next.setter
    def next(self, node):
        self._next = node


class QueueLinkedList:
    def __init__(self, list_size):
        self._front = None
        self._rear = None
        self._list_size = list_size

    def enqueue(self, value):
        node = Node(value)

        if self.is_empty():
            self._front = self._rear = node
        elif not (self.is_empty() or self.is_full()):
            self._rear.next = node
            self._rear = self._rear.next
            self._rear.previous = node.previous

    def dequeue(self):
        if self.is_empty():
            return None

        target = self._front
        self._front = self._front.next

        return target

    def peek(self):
        if self.is_empty():
            return None
        return self._front

    def is_empty(self):
        if not self._front:
            return True
        return False

    def is_full(self):
        count = 0
        current_node = self._rear
        while current_node:
            count += 1
            current_node = current_node.previous

        if count == self._list_size:
            return True
        return False

File: CPP/queue/queue.cpp
#include <iostream>
#include "queue.h"

using std::cout;
using std::endl;

template <typename T>
Node<T>::Node(const T input_value)
  : value(input_value)
{
  previous = nullptr;
  next = nullptr;
}

template <typename T>
void Node<T>::set_value(const T input_value)
{
  value = input_value;
}

template <typename T>
void Node<T>::set_previous(Node<T>* node)
{
  previous = node;
}

template <typename T>
void Node<T>::set_next(Node<T>* node)
{
  next = node;
}

template <typename T>
T const Node<T>::get_value() const
{
  return value;
}

template <typename T>
Node<T>* Node<T>::get_previous()
{
  return previous;
}

template <typename T>
Node<T>* Node<T>::get_next()
{
  return next;
}

template <typename T>
Node<T>::~Node()
{
  delete previous;
  delete next;
}

template <typename T>
QueueArray<T>::QueueArray()
  : front(-1), rear(-1), size(1)
{
  space = new T[1];
}

template <typename T>
void QueueArray<T>::Enqueue(const T input_value)
{
  if (IsFull())
    Resize();
  if (front == -1 && rear == -1)
  {
    space[++front] = input_value;
    ++rear;
  }
  else
  {
    space[++rear] = input_value;
  }
}

template <typename T>
T QueueArray<T>::Dequeue()
{
  T target = NULL;
  if (!IsEmpty())
  {
    target = space[front];
    space[front] = NULL;
    ++front;
  }
  return target;
}

template <typename T>
T QueueArray<T>::Peek()
{
  return space[front];
}

template <typename T>
void QueueArray<T>::Resize()
{
  int temp_size = size * 2;
  T* temp_array = new T[temp_size];
  for (int i = 0; i < size; ++i)
    temp_array[i] = space[i];
  space = temp_array;
  size = temp_size;
}

template <typename T>
bool QueueArray<T>::IsEmpty()
{
  if (front == rear)
    return true;
  return false;
}

template <typename T>
bool QueueArray<T>::IsFull()
{
  if (rear == size - 1)
    return true;
  return false;
}

template <typename T>
void QueueArray<T>::ShowArray()
{
  for (int i = front; i <= rear; ++i)
    cout << space[i] << endl;
}

template <typename T>
QueueArray<T>::~QueueArray()
{
  delete[] space;
}

template <typename T>
QueueList<T>::QueueList()
  : size(0)
{
  front = nullptr;
  rear = nullptr;
}

template <typename T>
void QueueList<T>::Enqueue(const T input_value)
{
  Node<T>* node = new Node<T>(input_value);
  if (front == nullptr)
  {
    front = node;
    rear = node;
  }
  else
  {
    Node<T>* current_node = front;
    while (current_node->get_next())
      current_node = current_node->get_next();
    current_node->set_next(node);
    node->set_previous(current_node);
    rear = node;
  }  
}

template <typename T>
Node<T>* QueueList<T>::Dequeue()
{
  Node<T>* target = nullptr;
  if (!IsEmpty())
  {
    Node<T>* current_node = front;
    Node<T>* next_node = current_node->get_next();
    target = current_node;
    current_node->set_next(nullptr);
    next_node->set_previous(nullptr);
    delete current_node;
    front = next_node;
  }
  return target;
}

template <typename T>
Node<T>* QueueList<T>::Peek()
{
  return front;
}

template <typename T>
bool QueueList<T>::IsEmpty()
{
  if (front == nullptr && rear == nullptr)
    return true;
  return false;
}

template <typename T>
void QueueList<T>::ShowList()
{
  Node<T>* current_node = front;
  while (current_node->get_next())
  {
    cout << current_node->get_value() << endl;
    current_node = current_node->get_next();
  }
  cout << current_node->get_value() << endl;
}

template <typename T>
QueueList<T>::~QueueList()
{
  delete front;
  delete rear;
}

int main()
{
  QueueArray<int>* queue_array = new QueueArray<int>();
  QueueList<int>* queue_list = new QueueList<int>();

  queue_array->Enqueue(5);
  queue_array->Enqueue(3);
  queue_array->Enqueue(8);
  queue_array->Enqueue(4);
  queue_array->Dequeue();
  queue_array->Enqueue(1);
  queue_array->Enqueue(7);
  queue_array->Dequeue();
  int peek01 = queue_array->Peek();
  cout << "Peek: " << peek01 << endl;
  queue_array->ShowArray();

  queue_list->Enqueue(5);
  queue_list->Enqueue(3);
  queue_list->Enqueue(8);
  queue_list->Enqueue(4);
  queue_list->Dequeue();
  queue_list->Enqueue(1);
  queue_list->Enqueue(7);
  queue_list->Dequeue();
  Node<int>* peek02 = queue_list->Peek();
  cout << "Peek: " << peek02->get_value() << endl;
  queue_list->ShowList();

  return 0;
}